harvardsettings_pi.php

```diff
@@ -299,8 +313,79 @@ public function import_users($arr) {
         return array(
             'message' => $message,
             'users_added' => $users_added,
-            'new_accounts_created' => $new_accounts_created
+            'new_accounts_created' => $new_accounts_created,
+            'unsuccessful' => $unsuccessful
+        );
+    }
+    
+    public function get_scalar_user(&$email_row, $huid_row, &$name, $unkown_emails) {
+        if ($email_row === '') {
+            if (!$unkown_emails || !isset($unkown_emails->$huid_row)) {
+                return "no identifiers";
+            }
+            $email_row = $unkown_emails->$huid_row->email;
+            $name = $name === '' ? $unkown_emails->$huid_row->name : $name;   
+        }
+        return $this->CI->users->get_by_email($email_row);
+    }
+    
+    public function new_scalar_user($email_row, $name, &$new_accounts_created) {
+        $new_user = array(
+            'email' => $email_row,
+            'fullname' => $name === '' ? 'placeholder' : $name,
+            'password' => 'DISABLEDPASSWORD',
+        );
+        if (!$this->CI->users->db->insert($this->CI->users->users_table, $new_user)) {
+            return false;
+        }
+        $user = $this->CI->users->get_by_email($email_row);
+        if ($user) {$new_accounts_created++;}
+        return $user;
+    }
+    
+    public function get_pds_emails($url) {
+        $c = curl_init();
+        curl_setopt($c, CURLOPT_URL, $this->CI->config->item('pds_base_url').$url);
+        curl_setopt($c, CURLOPT_RETURNTRANSFER, true);
+        curl_setopt(
+            $c,
+            CURLOPT_USERPWD,
+            $this->CI->config->item('pds_client_id') . ":" . $this->CI->config->item('pds_client_secret')
+        );
+        $output = json_decode(curl_exec($c));
+        curl_close($c);
+        
+        $return_obj = (object) array();
+        if (isset($output->message)) {
+            if ($output->message === "No valid key, huid, netid, or email found.") {
+                return $return_obj;
+            }
+        }
+        foreach($output as $person) {
+            $id = $person->univid;
+            $name = !$person->privacyFerpaStatus ? $person->names[0]->firstName . " " . $person->names[0]->lastName : 'placeholder';
+            $return_obj->$id = (object) array(
+                'name' => $name,
+                'email' => $person->loginName
+            );    
+        }
+        log_message(
+            'info',
+            $this->CI->data['login']->email." accessed the PDS for this query: ".$url
```

Comment:
I really like the logging here, but it occurs to me that the user could potentially change their email address on their profile, so we might want to add the user's scalar ID too (just in case).

Reply:
Totally, I'll add that in!